codec/decode.go

```diff
@@ -3,150 +3,170 @@ package codec
 // number of input bytes decoded, and err is nil
 // otherwise, it returns 0, 0, and error
-func DecodeInteger(b []byte) (o int64, bytesDecoded int64, err error) {
+func (sd *Decoder) DecodeInteger() (o int64, err error) {
+	b, err := sd.ReadByte()
+	if err != nil {
+		return 0, err
+	}
+
 	// check mode of encoding, stored at 2 least significant bits
-	mode := b[0] & 0x03
-	if mode == 0 { // 1 byte mode
-		return int64(b[0] >> 2), 1, nil
-	} else if mode == 1 { // 2 byte mode
-		o := binary.LittleEndian.Uint16(b) >> 2
-		return int64(o), 2, nil
-	} else if mode == 2 { // 4 byte mode
-		o := binary.LittleEndian.Uint32(b) >> 2
-		return int64(o), 4, nil
+	mode := b & 3
+	if mode <= 2 {
+		return sd.decodeSmallInt(b)
 	}
 
 	// >4 byte mode
-	topSixBits := (binary.LittleEndian.Uint16(b) & 0xff) >> 2
-	byteLen := topSixBits + 4
+	topSixBits := b >> 2
+	byteLen := int(topSixBits) + 4
 
-	if len(b) < int(byteLen)+1 {
-		err = errors.New("could not decode invalid integer")
+	buf := make([]byte, byteLen)
+	_, err = sd.reader.Read(buf)
+	if err != nil {
+		return 0, err
 	}
 
 	if err == nil {
-		bytesDecoded = int64(byteLen) + 1
 		if byteLen == 4 {
-			o = int64(binary.LittleEndian.Uint32(b[1:bytesDecoded]))
+			o = int64(binary.LittleEndian.Uint32(buf))
 		} else if byteLen > 4 && byteLen < 8 {
-			// need to pad upper bytes with zeros so that we can perform binary decoding
-			upperBytes := make([]byte, 8-byteLen)
-			upperBytes = append(b[5:bytesDecoded], upperBytes...)
-			upper := int64(binary.LittleEndian.Uint32(upperBytes)) << 32
-			lower := int64(binary.LittleEndian.Uint32(b[1:5]))
-			o = upper + lower
+			tmp := make([]byte, 8)
+			copy(tmp, buf)
+			o = int64(binary.LittleEndian.Uint64(tmp))
 		}
 
 		if o == 0 {
 			err = errors.New("could not decode invalid integer")
 		}
 	}
 
-	return o, bytesDecoded, err
+	return o, err
 }
 
 // DecodeBigInt decodes a SCALE encoded byte array into a *big.Int
 // Works for all integers, including ints > 2**64
-func DecodeBigInt(b []byte) (output *big.Int, bytesDecoded int64, err error) {
+func (sd *Decoder) DecodeBigInt() (output *big.Int, err error) {
+	b, err := sd.ReadByte()
+	if err != nil {
+		return nil, err
+	}
+
 	// check mode of encoding, stored at 2 least significant bits
-	mode := b[0] & 0x03
+	mode := b & 0x03
 	if mode <= 2 {
-		tmp, bytesDecoded, err := DecodeInteger(b)
-		return big.NewInt(tmp), bytesDecoded, err
+		tmp, err := sd.decodeSmallInt(b)
+		if err != nil {
+			return nil, err
+		}
+		return new(big.Int).SetInt64(tmp), nil
 	}
 
 	// >4 byte mode
-	topSixBits := (binary.LittleEndian.Uint16(b) & 0xff) >> 2
-	byteLen := topSixBits + 4
+	topSixBits := b >> 2
+	byteLen := int(topSixBits) + 4
 
-	if len(b) < int(byteLen)+1 {
-		err = errors.New("could not decode invalid integer")
-	}
-
-	// TODO: use io.Writer to return number of bytesDecoded
+	c := make([]byte, byteLen)
```

Comment:
still using `c`

Reply:
fixed, sorry about force push, accidentally commit trie stuff to this branch so i had to revert

Reply:
No change

Reply:
which line? looks updated?